### PR TITLE
[Clang Importer] Do not rely on being able to always import Foundation on-demand

### PR DESCRIPTION
**Pull request by artemcm**
When importing Clang types.
This is not an option with Explicit Module Builds. If the module being built does not (directly or transitively) depend on `Foundation`, then attempting to load `Foundation` will produce an error because implicit module loading is no longer allowed.

This change addresses a small number of cases where ClangImporter relies on being able to load `Foundation` on-demand:
- When importing a single Decl from a clang module, we check whether it has certain conformances by checking all extensions of the `NominalTypeDecl` of the `Decl` in question, to see if any of the extensions contain the conformance we are looking for, but we only check extensions whose parent module is either the original module of the `NominalTypeDecl` or the overlay module of the `NominalTypeDecl` or `Foundation`. It seems that we do not need to actually import `Foundation` here, just checking the module `Identifier` should be sufficient.
- In `maybeImportNSErrorOutParameter`, change the behavior to have an exit condition based on whether `Foundation` can be imported, before attempting to load it.
- When checking whether or not we are allowed to bridge an Objective-C type, it also looks sufficient to query whether or not `Foundation` *can* be imported, without loading it.

Resolves rdar://70381338

### CODE REVIEW

**Comment by artemcm** [comment]
@swift-ci please test

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@brentdax, @slavapestov, I don't have as deep of an understanding to know what I may be breaking here, really curious to hear what you think.

**Comment by artemcm** [comment]
@swift-ci please test

**Comment by artemcm** [comment]
@swift-ci Please Test Source Compatibility

**Comment by artemcm** [comment]
@swift-ci Please Test Source Compatibility

**Comment by artemcm** [comment]
@swift-ci Please Test Source Compatibility

**Comment by artemcm** [comment]
@swift-ci Please Test Source Compatibility

**Comment by artemcm** [comment]
@swift-ci Please Test Source Compatibility

**Comment by artemcm** [comment]
@swift-ci Please Test Source Compatibility Debug

**Comment by artemcm** [comment]
@swift-ci Please Test Source Compatibility Debug

**Comment by artemcm** [comment]
@swift-ci Please Test Source Compatibility Debug

**Review by beccadax — APPROVED** [review]
I *think* this will be okay, but I wouldn't be surprised if we see regressions in some weird edge cases where ClangImporter implicitly loading Foundation turns out to have been load-bearing. We'll need to keep an eye out for that.

File: lib/ClangImporter/ImportDecl.cpp [hunk 1]
```diff
@@ -5583,8 +5583,7 @@ namespace {
 }
 
 static bool conformsToProtocolInOriginalModule(NominalTypeDecl *nominal,
```
Comment (brentdax) [inline lib/ClangImporter/ImportDecl.cpp:5585]: Nit: Now that we're not passing in Foundation, this should probably be called something like `conformsToProtocolInFoundationModule()`.

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@swift-ci please smoke test Windows platform

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@swift-ci please smoke test macOS platform

**Comment by artemcm** [comment]
It looks like it wasn't load-bearing when I put up the PR, but became so since then for these tests:

```
Swift(macosx-x86_64) :: Interop/Cxx/class/constructors-objc-silgen.swift
Swift(macosx-x86_64) :: Interop/Cxx/class/constructors-objc-irgen.swift
```

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@swift-ci please smoke test macOS platform

**Comment by artemcm** [comment]
@swift-ci please smoke test macOS platform

**Comment by artemcm** [comment]
@swift-ci please smoke test macOS platform

**Comment by artemcm** [comment]
@swift-ci please smoke test macOS platform

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@swift-ci please smoke test Windows platform

**Comment by artemcm** [comment]
@swift-ci please smoke test Windows platform

**Pull request merged**